### PR TITLE
Split go-specific logic from shared logic

### PR DESCRIPTION
**Pull request by runesoerensen**
This PR builds on https://github.com/heroku/buildpacks-go/pull/244 and starts to split Go specific logic from shared logic by:

* Adding a new `common/inventory-utils` crate where shared functionality is moved to.
* Merging the Artifact struct’s `Version` and `SemanticVersion` fields into a generic field implementing the `Version` trait.
* Refactoring `Checksum` to separate business, domain, and serialization logic better. See https://github.com/heroku/buildpacks-go/pull/245/commits/63cb4c7743b58cecb35ef656c035388d20ded246, which includes most of the relevant changes related to that.

### CODE REVIEW

File: common/go-utils/src/vrs.rs [hunk 1]
```diff
@@ -57,75 +34,57 @@ impl VersionRequirement<GoVersion> for GoRequirement {
...
-pub enum VersionParseError {
+pub enum GoVersionParseError {
     #[error("Couldn't parse go version: {0}")]
     SemVer(#[from] semver::Error),
     #[error("Internal buildpack issue parsing go version regex: {0}")]
     Regex(#[from] regex::Error),
     #[error("Couldn't parse version. Unable to capture values from regex.")]
     Captures,
+    #[error("Couldn't parse go version: {0}")]
+    SemanticVersion(#[from] SemanticVersionParseError),
```
Comment (colincasey) [inline common/go-utils/src/vrs.rs:64]: Shouldn't this replace the `SemVer` error above? Are both required?

File: common/inventory-utils/src/semvrs.rs
Comment (colincasey) [inline common/inventory-utils/src/semvrs.rs:1]: Can you explain the purpose of this struct a bit more? The PR description alludes to why it was introduced but I'm not seeing the immediate value and I'm curious if I'm missing some key context. 

Is it purely to provide convenience to other consumers that may also have inventory items that abide by semver? If so, I'd say don't add this struct and leave that implementation up to each inventory implementation. The `Version` and `VersionRequirement` traits are easy enough to implement without this extra wrapper.

File: common/go-utils/src/vrs.rs [hunk 2]
```diff
@@ -139,27 +98,23 @@ impl GoVersion {
...
-impl From<GoVersion> for String {
-    fn from(ver: GoVersion) -> Self {
-        format!("{ver}")
-    }
-}
-
 impl fmt::Display for GoVersion {
     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        write!(f, "{}", self.0)
+        write!(f, "Go {}", self.0)
```
Comment (joshwlewis) [inline common/go-utils/src/vrs.rs:117]: Go versions and releases typically use the `go1.20.1` format upstream.

```suggestion
        write!(f, "go{}", self.0)
```

However, that's still not totally accurate. Prior to `go1.21.0`, the trailing `.0` wasn't included. If we want to use the historically accurate naming (I would prefer that), we'd need `1.20.0` to render as `go1.20` and `1.21.0` to render as `go1.21.0`.

**Review by joshwlewis — CHANGES_REQUESTED** [review]
It looks like the stringified version names have changed. It'd be ideal to respect the upstream naming.

File: common/go-utils/src/lib.rs [hunk 3]
```diff
@@ -1,3 +1,4 @@
-pub mod checksum;
+use toml as _;
```
[the remark this answers is not in the document]
Comment (runesoerensen) [reply common/go-utils/src/lib.rs:1]: Toml is actually used by the executables in the `bin` directory, so I don't think this is the right approach? At least doesn't make clippy happy :)

File: common/go-utils/src/vrs.rs (hunk 1 above)
Comment (runesoerensen) [reply common/go-utils/src/vrs.rs]: You're right -- this has been resolved by removing the `SemanticVersion` abstraction altogether :)

File: common/inventory-utils/src/semvrs.rs
Comment (runesoerensen) [reply common/inventory-utils/src/semvrs.rs:1]: Good point - I've addressed this by eliminating the `SemanticVersion` abstraction, and instead use traits to add the desired behavior directly on the `semver` `Version`/`VersionReq` implementations.

File: common/go-utils/src/vrs.rs (hunk 2 above)
Comment (runesoerensen) [reply common/go-utils/src/vrs.rs]: Agreed, this was a regression (which I attempted to justify here https://github.com/heroku/buildpacks-go/pull/245/commits/378b73ca045ed09b47dce45a9f7036e976216c02 ;-)) -- however, I definitely prefer displaying the actual version string over the parsed version.

The previous behavior was re-implemented [here](https://github.com/heroku/buildpacks-go/pull/245/commits/b0c757e75a972f9180c7d36de040783f63028261#diff-560d0a7bad62e4a300eba035429762f445bb85bb6ab9bc20f42b02c22545fb3cR96-R98), and the integration tests now verify this again https://github.com/heroku/buildpacks-go/pull/245/commits/a9bd12617724050031ec85cf9e70f5aa7c24c77a :)

**Comment by runesoerensen** [comment]
> It looks like the stringified version names have changed. It'd be ideal to respect the upstream naming.

Yes -- we're using the original version names again ([see above comment](https://github.com/heroku/buildpacks-go/pull/245#discussion_r1563747882), and the [`inventory.toml` is back to using the original version identifiers](https://github.com/heroku/buildpacks-go/pull/245/commits/0a667ea862528560068dc578744b6aa271d8e9f9)

File: common/go-utils/src/lib.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (colincasey) [reply common/go-utils/src/lib.rs:1]: Ah, my bad. I didn't notice it in the binary 😅

**Review by joshwlewis — APPROVED** [review]

**Pull request merged**